src/lib.rs

```diff
@@ -0,0 +1,564 @@
+    /// `resources/schema.sql`). This script is in principle idempotent so
+    /// running it twice should be safe, but it is also needless,
+    /// so this should only be requested on a first connection on startup.
+    pub async fn create(
+        config: &postgres::Config,
+        try_run_sql: Option<&str>,
+    ) -> anyhow::Result<Self> {
+        let mut client = DatabaseClient::create(config.clone(), postgres::NoTls).await?;
+
+        if let Some(sql) = try_run_sql {
+            client.as_ref().batch_execute(sql).await?
+        }
+
+        let prepared = P::prepare_all(&mut client).await?;
+
+        Ok(Self {
+            client,
+            prepared,
+        })
+    }
+}
+
+impl<P> AsRef<DatabaseClient> for DBConn<P> {
+    fn as_ref(&self) -> &DatabaseClient { &self.client }
+}
+
+impl<P> AsMut<DatabaseClient> for DBConn<P> {
+    fn as_mut(&mut self) -> &mut DatabaseClient { &mut self.client }
+}
+
+/// Holds information pertaining to block insertion into database.
+pub struct BlockInsertSuccess {
+    /// The time it took to insert the block.
+    pub time:         chrono::Duration,
+    /// The hash of the inserted block.
+    pub block_hash:   BlockHash,
+    /// The height of the inserted block.
+    pub block_height: AbsoluteBlockHeight,
+}
+
+/// A collection of possible errors that can happen while using the database.
+#[derive(Debug, Error)]
+pub enum DatabaseError {
+    /// Database error.
+    #[error("Error using the database {0}.")]
+    PostgresError(#[from] postgres::Error),
+    /// Other errors while processing database data.
+    #[error("Error using the database {0}.")]
+    OtherError(#[from] anyhow::Error),
+}
+
+/// Defines a set of necessary callbacks used by the database thread.
+#[async_trait]
+pub trait DatabaseHooks<D, P> {
+    /// Invoked by database thread every time a block has been received to be
+    /// inserted into the database.
+    async fn insert_into_db(
+        db_conn: &mut DBConn<P>,
+        data: &D,
+    ) -> Result<BlockInsertSuccess, DatabaseError>;
+
+    /// Invoked by the database thread to request the latest recorded height in
+    /// the database.
+    async fn on_request_max_height(
+        db: &DatabaseClient,
+    ) -> Result<Option<AbsoluteBlockHeight>, DatabaseError>;
+}
+
+/// A collection of possible errors that can happen while using the node to
+/// query data.
+#[derive(Debug, Error)]
+pub enum NodeError {
+    /// Error establishing connection.
+    #[error("Error connecting to the node {0}.")]
+    ConnectionError(tonic::transport::Error),
+    /// No finalization in some time.
+    #[error("Timeout.")]
+    Timeout,
+    /// Error establishing connection.
+    #[error("Error during query {0}.")]
+    NetworkError(#[from] v2::Status),
+    /// Query error.
+    #[error("Error querying the node {0}.")]
+    QueryError(#[from] v2::QueryError),
+    /// Query errors, etc.
+    #[error("Error querying the node {0}.")]
+    OtherError(#[from] anyhow::Error),
```

Comment:
Hmm.. In general I agree, but this is used in a way for users to easily feed an error with some added context from the hooks supplied to `run_service` (these hooks force users to return results with either `NodeError`s or `DatabaseError`s). See the `impl NodeHooks<TransactionLogData> for NodeDelegate` block in `main.rs` for an example.

What would you do in this case?

Reply:
I think I'll keep the anyhow version for now, as it is convenient and I think it makes sense while this is not released anywhere. But i strongly agree that if this was supposed to be released it would be bad to force onto the user.